### ISSUE TITLE
Mouseenter event not triggered when cursor moves from disabled button

### ISSUE THREAD

**Issue by stepancar** [facebook/react#10109]
BUG.

Mouseenter event not triggered when cursor moves from disabled button
see [example](https://alfa-laboratory.github.io/arui-feather/styleguide/#playground/code=%3Cdiv%3E%0A%20%0A%20%20%20%20%3Cdiv%20className='row'%3E%0A%20%20%20%20%20%20%20%20Hover%20on%20%20right%20button%20and%20then%20move%20cursor%20to%20left%20button%0A%20%20%20%20%20%20%20%20%3Cdiv%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cbutton%20onMouseEnter=%7B()=%3E%7Balert('Hey!');%7D%7D%20%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20Not%20Disabled%0A%20%20%20%20%20%20%20%20%20%20%20%20%3C/button%3E%0A%20%20%20%20%20%20%20%20%3C/div%3E%0A%20%20%20%20%20%20%20%20%3Cdiv%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cbutton%20disabled=%7B%20true%20%7D%3EDisabled%3C/button%3E%0A%20%20%20%20%20%20%20%20%3C/div%3E%0A%20%20%20%20%3C/div%3E%0A%0A%20%20%20%20%3Cdiv%20className='row'%3E%0A%20%20%20%20%20%20%20%20Hover%20on%20%20right%20button%20and%20then%20move%20cursor%20to%20left%20button%0A%20%20%20%20%20%20%20%20%3Cdiv%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cbutton%20onMouseEnter=%7B()=%3E%7Balert('Hey!');%7D%7D%20%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20Not%20Disabled%0A%20%20%20%20%20%20%20%20%20%20%20%20%3C/button%3E%0A%20%20%20%20%20%20%20%20%3C/div%3E%0A%20%20%20%20%20%20%20%20%3Cdiv%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cbutton%3ENot%20Disabled%3C/button%3E%0A%20%20%20%20%20%20%20%20%3C/div%3E%0A%20%20%20%20%3C/div%3E%0A%20%0A%3C/div%3E)

**What is the expected behavior?**
Should trigger Mouseenter event

React 15.5.3
Latest Chrome
MacOs 10.12.5

**Comment by Andarist** [comment]
**I think** its because *mouseout* event is not triggered for disabled form elements, according to `https://developer.mozilla.org/en/docs/Web/Events/mouseout` and React's synthetic mouseenter event works based on mouseout.

Aint sure if its possible to safely revert detection strategy to the mouseover (which would work), as there is probably some reason why its implemented like that.

**Comment by gaearon** [comment]
I’ll tag this as a bug, but not sure if it ever worked, or whether it’s possible to fix.

**Comment by Andarist** [comment]
What can be done about this temporary is to put smth like `margin-left: 2px;` on the disabled button. (1px doesnt seem to work).

**Comment by stepancar** [comment]
@gaearon could you check [this PR](https://github.com/facebook/react/pull/10149)?
Works good for this case

**Comment by aweary** [comment]
Verified this is still an issue in React 16: https://jsfiddle.net/ut5tqgbk/

**Comment by jquense** [comment]
Yes this is due to the way the polyfill is implemented and isn't really fixable with the current strategy since it relies on infering the paired event (enter/leave) from the relatedTarget of it's opposite. Just means that since disabled elements don't emit `mouseout` the entering element won't get a `mouseenter`

**Comment by NeekSandhu** [comment]
This is really a big trouble. I tried to get around it by attaching `onMouseEnter` to `li` tag wrapping the disabled `button`. Exhibited same behavior nonetheless 😞

**Comment by gaearon** [comment]
@jquense Does the fix in https://github.com/facebook/react/pull/10149 make sense to you?

**Comment by Justin-ZS** [comment]
Same problem here😞.(`react@16.3.2`)

**Comment by padraigfl** [comment]
Providing an update. Still seem to be having this issue on the latest versions of react, the mouseEnter event is attached to its direct parent in my case, tried putting it an extra level down (ie the mouse enter event is on the parent's parent) and still had the same issue.

Not a big deal in my case, can work around okay.

**Comment by esr360** [comment]
I can confirm this is still a bug.

We can see that in native HTML/JS, a non-disabled button will still trigger `mouseenter` when the previous target is a disabled button: https://codepen.io/esr360/pen/abvmoeJ?editors=1010 - the non-disabled button's background becomes blue, so we know it works.

In React, however, the same behaviour is not seen: https://codesandbox.io/s/throbbing-worker-227tz?file=/src/App.js - when we hover the non-disabled button _from_ the disabled button, `onMouseEnter` is not called and the background color does not change. If we hover the non-disabled button by any other means, we see the background color change to blue.

From an ex-CSS developer who believes CSS-in-JS is the future, please recognise the importance of fixing bugs like these. Thanks.

**Comment by mikeyamato** [comment]
Just to add some additional fodder, while `onMouseEnter` is not firing, `onMouseLeave` is. I'd like to use `onMouseEnter` vs. `onMouseOver` because it's not sent to any descendants. Super helpful when I have child components that I don't want triggering events.

Any good work arounds in the time being?

**Comment by snowystinger** [comment]
This would be useful when implementing a SpinButton component. When hitting the max or min value, the corresponding button would be disabled. Trying to implement hover state on mouse enter for the opposing direction can be accomplished with mousemove, but it'd make more sense if it worked with mouseenter.
For example:
![Screen Shot 2020-12-02 at 5 33 32 PM](https://user-images.githubusercontent.com/698229/100952118-c2b7ec80-34c4-11eb-8327-e621cfeec081.png)

**Comment by vlazh** [comment]
It's still not fixed in 17.0.2 - [example](https://codesandbox.io/s/react-onmouseenter-on-disabled-buttons-forked-n5boc?file=/src/index.js)

**Comment by benhall847** [comment]
I can also confirm this is still not fixed

**Comment by markovd18** [comment]
Any updates on this? This is still an issue in React 18. For us neither `onMouseEnter` nor `onMouseLeave` events are fired on disabled button.

**Comment by purahan** [comment]
**Re-tested this on current browsers and React. The original symptom doesn't reproduce anymore, but there's a closely related issue still worth flagging.**

Setup: two sibling `<button>`s, one disabled, flush against each other with no gap, single continuous cursor drag across the shared edge. Tested with:
- Raw native DOM events (`mouseover`/`mouseout`/`pointerover`/`pointerout`), capture-phase listeners on `document`
- Real React 18 `onMouseEnter`/`onMouseLeave` handlers

Current Chrome and current Firefox both fire native hover events cleanly across this boundary. `mouseout`/`mouseleave` on the disabled button pairs almost instantly with `mouseover`/`mouseenter` on the sibling, every time, across dozens of repeated crossings. React 18 passes this through correctly too. So the original report here, moving off a disabled button failing to trigger enter on the next element, doesn't reproduce for me on either engine in 2026. Haven't tested Safari/WebKit, so if anyone can check there that'd help close the gap.

While digging into this though, I traced a separate, still-active behavior that might actually be what some later comments on this thread were running into. `onMouseEnter` is explicitly suppressed on the disabled element itself, in `shouldPreventMouseEvent` in `getListener.js`. That's intentional, added in #17675 back in Feb 2020, but that PR's own discussion has a reviewer raising almost the exact use case several people here have mentioned: showing a tooltip or help text explaining why a button is disabled, which needs hover feedback on the disabled element itself. The PR author also left a comment mid-review noting that `onMouseLeave`/`onMouseExit` doesn't have this problem and wondering if the fix actually belonged in event registration rather than this suppression list. That thread doesn't seem to have been followed up on.

Given that the original browser-inconsistency assumption this issue was built on doesn't seem to hold anymore, and #17675 shipped a targeted fix over an objection that was never really resolved, would a maintainer have thoughts on revisiting whether `onMouseEnter` belongs in that switch statement at all? Happy to put together a PR removing it (plus the accompanying test in `ReactBrowserEventEmitter-test.internal.js`) and adding a regression test for the tooltip-on-disabled-button case, if that direction seems reasonable. Didn't want to open one blind without some signal first given how old and layered this area is.
